### PR TITLE
xonsh shell binding

### LINKED ISSUE #2009: Xonsh Support
Trying to implement support for the Xonsh shell.

### PR DESCRIPTION
**Pull request by anki-code**
Added the link to xonsh shell binding.

Closes #2009

### CODE REVIEW

**Comment by PH111P** [comment]
I'm don't really like the fact that this PR essentially just adds an advertisement to an external project:
* If a user desires a powerline for xonsh, which the phrasing in this PR, they may think that just installing powerline would be enough—which is not the case.
* We provide no support for external code, however, adding a link to external code may wrongly give that impression.
* In the case that the external code causes bugs or other unwanted behavior, a user may blame powerline, while in fact the issues were caused by third-party code.

TL;DR: I won't add advertisements to third-party projects; however, I could imagine merging said third-party code directly, if it is well written and comes with the necessary tests.

**Comment by anki-code** [comment]
Hello @PH111P. You're points is reasonable but mostly your conclusions are wrong.

> If a user desires a powerline for xonsh, which the phrasing in this PR, they may think that just installing powerline would be enough—which is not the case.

To solve this I can delete xonsh mention from the first line. And leave only "Third-party bindings" section.

> We provide no support for external code, however, adding a link to external code may wrongly give that impression.

No. There is idiom - [Third-party software](https://techterms.com/definition/thirdparty): 

    In computer programming, a third-party software component is a software component 
    developed by an entity other than the original vendor of the development platform. 

And I point it out in the section title - "Third-party bindings".

> In the case that the external code causes bugs or other unwanted behavior, a user may blame powerline, while in fact the issues were caused by third-party code.

No. It's a link to third party solution.

> I'm don't really like the fact that this PR essentially just adds an advertisement to an external project:

Are you strongly believe that "advertisement" word is appropriate using in Open Source community? The developer made powerline support on new platform and close the feature request #2009 and then I suggest just add the mention of the project in the "Third-party" section at the bottom of the readme. Are you serious think that it's have a bad consequences for the powerline project? 

It's not an "advertisement". It's a mention of new platform where powerline could be used.

-- 

Meanwhile I removed "xonsh" in the first line and leave only "Third-party" section at the bottom of the page. Could you please [review this again](https://github.com/anki-code/powerline/blob/patch-1/README.rst) or your response is strongly NO?

**Comment by PH111P** [comment]
My response is a strong “no” indeed, let me elaborate some more:

Placing a link to third-party software in a prominent place in the README file creates the impression that powerline is to some extent related to or supports said third-party software—which powerline isn't, apart from said third-party software using powerline as a dependency (in this sense, such a link is just an advertisement for unrelated/external code). In particular, adding such a link would require some maintainer to constantly watch said third-party software to ensure linking to it is still justified / said third-party software is still maintained, compatible, non-malicious, etc.

In general, adding a “third-party” section seems wrong to me—how do we choose which external projects to include and which not? Who checks these third-party projects / keeps links to them up-to-date, etc?

Please note that I am not opposed to the feature that the third-party software provides in this case, however, for inclusion in powerline, said feature should be added directly via a PR (and have appropriate accompanying tests).

**Comment by anki-code** [comment]
Logic is valid but too much perfectionism. In real life there is no needed to check and double check that third party works. It's contradicts with Github idea of "forking" everywhere any time. But it make opportunity the users to know something new i.e. that third party bindings are exist and this is the part of universe. 

If I ask you "Could you write binding for <...>?" you tell me also "No". As result your position is "a man in the middle": say "No" to external contributions and say "No" to personal contribution. You wait for someone who gives you ideal solution to merge it. Okay.

I have no any wish to give you any other PRs because I'll face with additional bag of perfectionism. It's not for me, sorry.

I recommend you to google the second new word after "Third-party" - [Ecosystem](https://en.wikipedia.org/wiki/Ecosystem).

**Comment by PH111P** [comment]
It's sad to hear that you don't want your work to be merged upstream. 

I'd like to point out that I'm not against _contributions_ to powerline. However, as powerline has many users, a certain level of quality is indeed required for such contributions. (Please understand that this is also so that the code, and the project, remain maintainable.)  Linking to external third-party code, powerline would have no way to ensure a certain level of quality, hence I'm strongly against it. 

Please observe the qualitative difference between “I want you to link to my project” and “I want you to integrate my project into your code; (optionally: I have made the following efforts already to make sure that it works properly and does so in the future as well)”. Please also note that when opening a PR, by no means it needs to be complete and that you can get support in fixing some issues you may encounter when implementing a new feature or while trying to integrate your new feature to the code base.

Finally, I'd advise you to look through the merged PR's to get an idea of the many great contributions to powerline by many different people—As I said earlier, I am by no means against merging cool new features (and certainly, the work you linked to implements a cool new feature).

**Comment by anki-code** [comment]
Got it! Thank you for your attention, detailed answers and great project!
